### ISSUE TITLE
30 concurrent agents + each of them 30 overviews - script/scheduler.rb takes 100% CPU time - background jobs cannot be processed

### ISSUE THREAD

**Issue by martini** [zammad/zammad#2108]
### Infos:

* Used Zammad version: 2.5
* Installation method (source, package, ..): any
* Operating system: any
* Database + version: any
* Elasticsearch version: any
* Browser + version: any

### Expected behavior:

* Work well with more then 30 concurrent agents + each ob them 30 overviews

### Actual behavior:

* background jobs cannot be processed - load for script/scheduler.rb is to high

**Comment by siepkes** [comment]
We are also experiencing reliability issues with the Zammad scheduler. We would like to improve the scheduler to be more resilient. This especially hampers the ability to send outbound mail which (IMHO) must never get blocked by other things the scheduler is doing.

We've done some small work on making the scheduler more reliable. We haven't fully delved in to it but I was wondering if "outsourcing" this background job processing to a more heavy weight specialized queueing solution like sidekiq (just an example) which allows for prioritization, monitoring, etc. be a good fit?

**Comment by martini** [comment]
@siepkes I think we have two things. 1) Improve performance/CPU usage of 'Session.jobs' and second switch to active jobs to allow to change the background job processing (e. g. by using sidekiq (just an example). What do you think?

PS: Pull requests are alway welcome.

**Comment by muhammadn** [comment]
I had some work in porting delayed job to sidekiq for zammad. Basically database backed queueing like delayed job will have scaling limitations.

For sidekiq i purpose saving the job ids (to manage the jobs)  into the database and let  other queuing jobs like email or triggers done by sidekiq.+ redis.

If @martini  approves it then i will port the background jobs code to sidekiq but the question remains if we should use ActiveJob (or call sidekiq directly).

@martini i can work on this since i already have that  branch.

**Comment by monotek** [comment]
https://github.com/mperham/sidekiq/wiki/Using-Redis#multiple-redis-instances

I don't like this at all. Redis persistance would create another stateful part in zammad :-/

**Comment by siepkes** [comment]
I agree that those are 2 separate problems; The session job taking too long is a scalability issue which is something that needs to be addressed in order to make Zammad scale well to say 50 or 100 concurrent users but is different from the scheduler resiliency issue.

As for the lower level implementation details on how to incorporate background job processing I (currently) do not have enough knowledge of Zammad yet to delve into too much details. I do however have experience with implementing queueing jobs, messages, etc. and I can share my opinion on the high level implementation (Geez Jasper, you have an opinion on something? Thats a shocker... :smile: ). 

From what I know about sidekiq (though I could be wrong) is that there is no way for a worker to "acknowledge" a job. That means that if a worker process dies and it has grabbed a job from the queue that job simply vanishes. This is why often solutions like RabbitMQ are used; Jobs are put in a queue, processed by workers who acknowledge to RabbitMQ they have successfully completed the job. There is also an added bonus here; You can run multiple workers, even on different machines.

If you really don't want to use solutions like RabbitMQ, Sidekiq, etc. you could write your own queue which uses PostgreSQL. Though I guess that will lead to solving a lot of problems these frameworks have already solved (queue metrics, qos, etc) so personally I don't think that a road you want to go.

Also in the this cloud (native) era failure is the new normal. For example in our case Zammad connects to a PostgreSQL cluster (Stolon). There are times when there is a switch / fail -over during which the DB is unavailable for a couple of seconds and connections gets reset. Applications should be able to handle those conditions gracefully. Usually just retry until other services it needs come back up.

**Comment by thorsteneckel** [comment]
I think @martini opened this issue to cover the first point in his list.
But just give you our current opinion on the second point:
When refactoring the job processing in general we will probably use [Rails Active Job](http://edgeguides.rubyonrails.org/active_job_basics.html) which provides a single abstraction interface for using a [variation of backends](https://github.com/rails/rails/tree/master/activejob/lib/active_job/queue_adapters). There should be one adapter for everyone depending on the use case.
Our main course is to keep the setup of Zammad as simple as possible but don't block the way for more advanced use cases as e.g. @monotek stated.
Therefore we won't replace the hardcoded Delayed::Job implementation with a hardcoded Sidekiq /Redis implementation @muhammadn .

**Comment by muhammadn** [comment]
@thorsteneckel Thanks for the clarification! 🤓I also have a branch that implements ActiveJjob.  Is that something that i should pursue as well?

But i saw some code that uses Delayed::Job that queries the database for the background job records and process it as necessary (example: deleting a job)

This is the tricky part because even if we port over fully to ActiveJob, we cannot actively query using DelayedJob to manage it (you can partially port it, leaving DelayedJob to query but it won't be portable to ActiveJob + sidekiq/resque, etc) . So my idea is to use Job IDs instead which ActiveJob provides which will work with delayed_job/sidekiq or resque. 

The hardcoded part won't be Sidekiq in this case, but most of the background job code will be replaced with ActiveJob.

**Comment by siepkes** [comment]
@thorsteneckel Your right, this thread has drifted a bit away from it's original problem.

Would it be possible to make the delayed job mechanism pluggable so people can write their own? 

I think our objectives are somewhat at odds with each other; You guys want to keep things simple (which I can understand) and we try to make things more stable and predictable for a more advanced, larger installation (which sometimes increases complexity). This is basically the same discussion (don't mean that in a negative way) we had in #806. We ended up creating the REST interface in our internal Zammad fork which works really well for us; We never missed an inbound mail again (even when the scheduler hangs). Also we don't encounter issues like #2092 since the REST API tells Postfix to bounce it if Zammad doesn't want the message. So from our perspective this implementation made total sense. However it does increase complexity so I can understand that the implementation is at odds with your objective. And I think this is basically a similar situation.

**Comment by thorsteneckel** [comment]
IMHO replacing Delayed::Jobs with Active Jobs is a (huge) task that needs proper evaluation and specification of current and upcoming use cases. We need to provide a solid solution. Surely there will be some trade offs down the road that we can't avoid, too.
However, we don't have the complete picture now but just a rough idea where we want to go so I can't answer your questions. We are currently focusing on building the knowledge base but will tackle this topic probably by end of the year, ask for your use cases and feedback, create a plan from it and build this baby. Help and feedback is always welcome! But for now I think it's way to early to start in any direction.
So let's get back to topic and see how we can improve the `Session.jobs` without changing the whole construct 🚀

**Comment by muhammadn** [comment]
Yeah. Sessions.job is totally a different issue totally, basically running in a loop and blocking delayed_job. I also had looked into the forums and couldn't find a place to talk about this.

Should i open a new issue or should i discuss it in the forums? 😬

@siepkes FYI. you can take a look at my current progress of porting Zammad to use ActiveJob. There are 3 files in the source code that queries for things like rescheduling by querying Delayed Job table directly.

app/models/scheduler.rb
controllers/monitoring_controller.rb
models/channel/driver/twitter.rb

So basically i had moved all the background process code to app/job and all functionalities work except import which might take some time i think.

All you need is to swap :sidekiq to :delayed_job and vice versa in config/application.rb and it just works as usual. However do note that for sidekiq you need to start the worker `bundle exec sidekiq` last (after rails, scheduler, etc).

My branch is here:
https://github.com/muhammadn/zammad/tree/activejob

<img width="1392" alt="screen shot 2018-07-06 at 1 21 34 am" src="https://user-images.githubusercontent.com/3349747/42338774-81fbe02e-80bd-11e8-941a-9eeeb208c719.png">
<img width="1392" alt="screen shot 2018-07-06 at 1 22 50 am" src="https://user-images.githubusercontent.com/3349747/42338775-8249e864-80bd-11e8-96a3-e3c0934f1997.png">

**Comment by thorsteneckel** [comment]
@muhammadn - yes, please stick to the [community board](http://community.zammad.org/) for now. Thanks!
Please note that the ActiveJob branch is highly experimental without the required test coverage to go into production. Additionally there are some other functionalities that need get covered, too. We will address those later, as described in my earlier comment.

**Comment by saschajullmann** [comment]
Hello,
I was asked to post some info here because I opened up an issue in the docker-compose repo but @thorsteneckel thinks it has nothing to do with docker and rather is an application bug.

The ticket I'm referring to is:    https://github.com/zammad/zammad-docker-compose/issues/82

In Short: We run zammad in production with the docker-compose setup. We have around 115K tickets, 10 agents and a machine with 4 cores and 8gig of ram.

The problem is very high memory consumption up to a point where I have to manually restart the scheduler container (otherwise it would run full and I'd have to manually restart the EC2 instance).

At this point I have to restart the scheduler container every so often (will probably add a cron job for this since a solution is not really in sight) so that the system doesn't crash.

I don't know what the scheduler does and I wish I could provide more insight but I'm rather clueless.

**Comment by thorsteneckel** [comment]
Hi @saschajullmann - nice to see you again ;) How many Overviews do you have in your system?

**Comment by siepkes** [comment]
@saschajullmann @thorsteneckel We also wanted more insight in what the scheduler was doing so we (recently) added a Prometheus endpoint to it. This allows us to gather metrics on queue size, errors, etc. As a bonus you can also use the Prometheus alert manager if things really go of the rails. Would you guys be interested in a PR for that?

**Comment by thorsteneckel** [comment]
Hi @siepkes - good to read you again! Since it's a bit off topic I'd like to ask what @monotek our docker expert thinks about it.

**Comment by siepkes** [comment]
@thorsteneckel Sorry for keep taking this topic off-topic :wink: However what we implemented is in the Ruby scheduler and therefor not really related to the docker image. Basically what it does is start a small (configurable) HTTP server when the scheduler script is started and expose a Prometheus compatible API / endpoint. 

Prometheus expects to pull / scrape metrics (from an HTTP endpoint) at a fixed intervals instead of getting metrics pushed to it (so thats the opposite of metric gathering solutions like statsd, graphite, etc.).

**Comment by saschajullmann** [comment]
@thorsteneckel we currently have 5 overviews which are more or less the standard ones plus one overview for high priority tickets.

We also had a "ALL Tickets" overview for admins just to see how many tickets got imported. I deleted this overview in order to see whether this would help the issue but I can't say that it did.

@siepkes the Prometheus endpoint sounds interesting. While we don't have a prometheus server yet, I toyed with the idea of setting one up in order to gather system metrics.

**Comment by martini** [comment]
I just pushed a commit with significant performance improvement to stable (without any background job changes). It would be nice of somebody could verify it if `script/scheduler.rb` takes then less CPU time (not, usage of the CPU should significant go down after about 60 sec.). Thanks for feedback!

Thanks to @thorsteneckel !
